crates/ty_python_semantic/src/semantic_index/symbol.rs

```diff
@@ -123,6 +124,26 @@ pub(super) struct SymbolTable {
     ///
     /// Uses a hash table to avoid storing the name twice.
     map: hashbrown::HashTable<ScopedSymbolId>,
+
+    // The resolutions of variables that are either used-but-not-defined or explicitly marked
+    // `global` or `nonlocal` in this scope. These (keys) are similar to what CPython calls "free"
+    // variables, except that we also include variables marked `global`.
+    pub(super) definitions_in_enclosing_scopes:
+        FxHashMap<ScopedSymbolId, (FileScopeId, ScopedSymbolId)>,
+
+    // The inverse of `definitions_in_enclosing_scopes` above: variables defined in this scope,
+    // and not marked `global` or `nonlocal` here, which are used or bound in nested scopes.
+    // These (keys) are similar to what CPython calls "cell" variables, except that this scope
+    // may also be the global scope. We also include nested scopes that only mark a variable
+    // `global` or `nonlocal` but don't touch it after that. Those aren't very useful, but IDE
+    // features like "rename all" will still want to know about them.
+    pub(super) references_in_nested_scopes:
+        FxHashMap<ScopedSymbolId, Vec<(FileScopeId, ScopedSymbolId)>>,
```

Comment:
You Could use a `SmallVec` here. This would allow us to store 4 scoped symbol ids inline.